cnitool/main.go

```diff
@@ -0,0 +1,267 @@
+	if cniBinDir == "" {
+		return fmt.Errorf("--cni-bin-dir is required")
+	}
+	cniBinDirs = filepath.SplitList(cniBinDir)
+
+	if cniConfDir == "" {
+		cniConfDir = os.Getenv(EnvNetDir)
+	}
+	if cniConfDir == "" {
+		return fmt.Errorf("--cni-conf-dir is required")
+	}
+
+	if capabilityArgs == "" {
+		capabilityArgs = os.Getenv(EnvCapabilityArgs)
+	}
+	if cniArgs == "" {
+		cniArgs = os.Getenv(EnvCNIArgs)
+	}
+	if i := os.Getenv(EnvCNIIfname); i != "" {
+		ifName = i
+	}
+
+	if len(capabilityArgs) > 0 {
+		if err := json.Unmarshal([]byte(capabilityArgs), &capabilityArgs); err != nil {
+			return fmt.Errorf("failed to parse capability args: %w", err)
+		}
+	}
+
+	if len(cniArgs) > 0 {
+		for _, pair := range strings.Split(cniArgs, ";") {
+			kv := strings.Split(pair, "=")
+			if len(kv) != 2 || kv[0] == "" || kv[1] == "" {
+				return fmt.Errorf("invalid cni-args pair %q", pair)
+			}
+			cniArgsParsed = append(cniArgsParsed, [2]string{kv[0], kv[1]})
+		}
+	}
+
+	return nil
+}
+
+// doAttach executes either add, del, or check
+func doAttach(ctx context.Context, op string, args []string) error {
+	if err := loadArgs(); err != nil {
+		return err
+	}
+
+	if len(args) != 2 {
+		return fmt.Errorf("2 arguments required")
+	}
+
+	name := args[0]
+	netns := args[1]
+
+	cninet := libcni.NewCNIConfig(cniBinDirs, nil)
+
+	rt := &libcni.RuntimeConf{
+		ContainerID:    containerID(netns),
+		NetNS:          netns,
+		IfName:         ifName,
+		Args:           cniArgsParsed,
+		CapabilityArgs: capabilityArgsParsed,
+	}
+
+	netconf, err := libcni.LoadConfList(cniConfDir, name)
+	if err != nil {
+		return err
+	}
+
+	switch op {
+	case "add":
+		result, err := cninet.AddNetworkList(ctx, netconf, rt)
+		if result != nil {
+			_ = result.Print()
+		}
+		return err
+	case "del":
+		return cninet.DelNetworkList(ctx, netconf, rt)
+	case "check":
+		return cninet.CheckNetworkList(ctx, netconf, rt)
+	}
+
+	return nil
+}
+
+func doStatus(cmd *cobra.Command, args []string) error {
+	if err := loadArgs(); err != nil {
+		return err
+	}
+
+	if len(args) != 1 {
+		return fmt.Errorf("1 argument required")
+	}
+	name := args[0]
+
+	cninet := libcni.NewCNIConfig(cniBinDirs, nil)
+	netconf, err := libcni.LoadConfList(cniConfDir, name)
+	if err != nil {
+		return err
+	}
+
+	err = cninet.GetStatusNetworkList(cmd.Context(), netconf)
+	if err != nil {
+		return fmt.Errorf("network %s is not ready: %w", name, err)
+	}
+	cmd.Printf("Network %s is ready for ADD requests\n", name)
+	return nil
+}
+
+func doGC(cmd *cobra.Command, args []string) error {
+	if err := loadArgs(); err != nil {
+		return err
+	}
+
+	if len(args) < 1 {
+		return fmt.Errorf("1 argument required")
+	}
+
+	validAttachments := []types.GCAttachment{}
+	for _, netns := range args[1:] {
+		pair := strings.Split(netns, ":")
+		ifname := "eth0"
+		if len(pair) > 1 {
+			ifname = pair[1]
+		}
+
+		validAttachments = append(validAttachments, types.GCAttachment{
+			IfName:      ifname,
+			ContainerID: containerID(netns),
```

Comment:
this is wrong AFAIU - you're checking the containerID based on the "full" netns value, which can include the name of the interface.